fuel-derive/src/serialize.rs

```diff
@@ -0,0 +1,551 @@
+            if should_skip_field_binding(binding) {
+                quote! {}
+            } else {
+                quote! {
+                    ::fuel_types::canonical::Serialize::encode_dynamic(#binding, buffer)?;
+                }
+            }
+        });
+        quote! {
+            #encode_dynamic_iter
+        }
+    });
+
+    // Handle #[canonical(serialize_with = function)]
+    let data_helper = attrs.serialize_with.as_ref();
+    let size_helper_static = attrs.serialized_size_static_with.as_ref();
+    let size_helper_dynamic = attrs.serialized_size_dynamic_with.as_ref();
+    if let (Some(data_helper), Some(size_helper_static), Some(size_helper_dynamic)) =
+        (data_helper, size_helper_static, size_helper_dynamic)
+    {
+        let size_no_dynamic = attrs
+            .SIZE_NO_DYNAMIC
+            .expect("serialize_with requires SIZE_NO_DYNAMIC key");
+
+        return s.gen_impl(quote! {
+            gen impl ::fuel_types::canonical::Serialize for @Self {
+                const SIZE_NO_DYNAMIC: bool = #size_no_dynamic;
+
+                #[inline(always)]
+                fn encode_static<O: ::fuel_types::canonical::Output + ?Sized>(&self, buffer: &mut O) -> ::core::result::Result<(), ::fuel_types::canonical::Error> {
+                    #data_helper(self, buffer)
+                }
+
+                fn encode_dynamic<O: ::fuel_types::canonical::Output + ?Sized>(&self, buffer: &mut O) -> ::core::result::Result<(), ::fuel_types::canonical::Error> {
+                    ::core::result::Result::Ok(())
+                }
+
+                fn size_dynamic(&self) -> usize {
+                    #size_helper_dynamic(self)
+                }
+            }
+
+            gen impl ::fuel_types::canonical::SerializedSize for @Self {
+                #[inline(always)]
+                fn size_static(&self) -> usize {
+                    #size_helper_static(self)
+                }
+            }
+        })
+    } else if ![
+        data_helper.is_none(),
+        size_helper_static.is_none(),
+        size_helper_dynamic.is_none(),
+    ]
+    .into_iter()
+    .all(|v| v)
+    {
+        panic!("serialize_with, serialized_size_static_with and serialized_size_dynamic_with must be used together");
+    }
+
+    let (variant_size_static, size_no_dynamic): (Vec<TokenStream2>, Vec<TokenStream2>) =
+        s.variants()
+            .iter()
+            .map(|v| constsize_fields(v.ast().fields))
+            .unzip();
+
+    let size_no_dynamic = size_no_dynamic.iter().fold(quote! { true }, |acc, item| {
+        quote! {
+            #acc && #item
+        }
+    });
+
+    // #[repr(int)] types have a known static size
+    let impl_size = if let Some(repr) = repr {
+        // Repr size, already aligned up
+        let repr_size: usize = match repr.as_str() {
+            "u8" => 8,
+            "u16" => 8,
+            "u32" => 8,
+            "u64" => 8,
+            "u128" => 16,
```

Comment:
It looks like every unsigned integer type is going to be serialized to take up 8 bytes (except 16 for `u128`) - is that correct? Is this part of the specification? 
Maybe using constants here, with a mention of the spec, would make this a little more clear and a little more flexible in case the specification ever changes.